tests/unit/json_test.c

```diff
@@ -1908,116 +1928,106 @@ static void test_json_null_not_null(void)
+    JsonElement *extra = NULL;
+    JsonElement *expected = NULL;
+    char *expected_string = NULL;
+    char *actual_string = NULL;
+
     if (JsonParse(&base_raw, &base) != JSON_PARSE_OK)
     {
-        return false;
+        printf("error: could not parse base: %s\n", base_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *extra = NULL;
     if (JsonParse(&extra_raw, &extra) != JSON_PARSE_OK)
     {
-        return false;
+        printf("error: could not parse extra: %s\n", extra_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *expected = NULL;
     if (JsonParse(&expected_raw, &expected) != JSON_PARSE_OK) {
-        return false;
+        printf("error: could not parse expected: %s\n", expected_raw);
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *actual = JsonObjectMergeDeep(base, extra);
+    expected_string = JsonToString(expected);
+
+    actual = JsonObjectMergeDeep(base, extra);
     if (actual == NULL || actual == base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    /* JsonCompare is naive and does not handle containers properly, thus we
-     * need a custom comparison function to test for equality. */
-    if (!compare_json_object_merge_deep(actual, expected))
+    actual_string = JsonToString(actual);
+
+    if (JsonCompare(actual, expected) != 0)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonDestroy(actual);
+    DESTROY_AND_NULL(JsonDestroy,actual);
     actual = JsonObjectMergeDeepInplace(base, extra);
+
+    // JsonObjectMergeDeepInplace() returns the first parameter (base) so actual should point to base
     if (actual != base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    if (!compare_json_object_merge_deep(actual, expected))
+    if (JsonCompare(actual, expected))
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonDestroy(base);
-    JsonDestroy(extra);
-    JsonDestroy(expected);
+    finish:
+        if (actual_string)
+        {
+            if (pass == false)
+            {
+                printf("merged json  : %s\n", actual_string);
+            }
+            free(actual_string);
+        }
+        if (expected_string)
+        {
+            if (pass == false)
+            {
+                printf("expected json: %s\n", expected_string);
+            }
+            free(expected_string);
+        }
+        // in second step of JsonObjectMergeDeepInplace() actual should point to base
+        if (actual != base && actual != NULL)
+        {
+            JsonDestroy(actual);
+        }
+        if (base != NULL)
+        {
+            JsonDestroy(base);
+        }
+        if (extra != NULL)
+        {
+            JsonDestroy(extra);
+        }
+        if (expected != NULL)
+        {
+            JsonDestroy(expected);
+        }
```

Comment:
Still think these checks for `!= NULL` should not be necessary, but they don't hurt.

Reply:
I agree, also since as a reader it makes it look like it's necessary to check for `NULL`. To be extra clear, you could do:

```
JsonDestroy(expected); // Can be NULL / no-op
```

Reply:
yep, changed. thanks!